services/cli/src/utils/backtest_rebalance.rs

```diff
@@ -0,0 +1,278 @@
+use crate::error::CliError;
+use futures::future::try_join_all;
+use jito_steward::{Config, constants::TVC_ACTIVATION_EPOCH, score::validator_score};
+use solana_sdk::native_token::LAMPORTS_PER_SOL;
+use sqlx::{Pool, Postgres};
+use stakenet_simulator_db::{
+    cluster_history::ClusterHistory, cluster_history_entry::ClusterHistoryEntry,
+    epoch_rewards::EpochRewards, validator_history::ValidatorHistory,
+    validator_history_entry::ValidatorHistoryEntry,
+};
+use std::collections::HashMap;
+use std::sync::Arc;
+use tracing::{error, info};
+use validator_history::ClusterHistory as JitoClusterHistory;
+
+#[derive(Clone, Debug)]
+pub struct RebalancingCycle {
+    pub start_epoch: u16,
+    pub end_epoch: u16,
+    pub selected_validators: Vec<String>,
+    pub starting_total_lamports: u64,
+    pub ending_total_lamports: u64,
+    pub starting_stake_per_validator: u64,
+}
+
+pub async fn rebalancing_simulation(
+    db_connection: &Pool<Postgres>,
+    steward_config: &Config,
+    simulation_start_epoch: u16,
+    simulation_end_epoch: u16,
+    steward_cycle_rate: u16,
+    number_of_validator_delegations: usize,
+) -> Result<Vec<RebalancingCycle>, CliError> {
+    let mut rebalancing_cycles = Vec::new();
+    let mut current_cycle_start = simulation_start_epoch;
+
+    let mut lamports_after_staking = LAMPORTS_PER_SOL
+        .checked_mul(number_of_validator_delegations as u64)
+        .ok_or(CliError::ArithmeticError)?;
+
+    let histories = ValidatorHistory::fetch_all(db_connection).await?;
+    let cluster_history = ClusterHistory::fetch(db_connection).await?;
+    let cluster_history_entries = ClusterHistoryEntry::fetch_all(db_connection).await?;
+    let jito_cluster_history =
+        cluster_history.convert_to_jito_cluster_history(cluster_history_entries);
+
+    let jito_cluster_history = Arc::new(jito_cluster_history);
+
+    info!("Fetching all validator history entries...");
+    let all_entries =
+        ValidatorHistoryEntry::fetch_all_validator_history_entries(db_connection).await?;
+
+    let mut entries_by_validator: HashMap<String, Vec<ValidatorHistoryEntry>> = HashMap::new();
+    for entry in all_entries {
+        entries_by_validator
+            .entry(entry.vote_pubkey.clone())
+            .or_insert_with(Vec::new)
+            .push(entry);
+    }
+
+    let entries_by_validator = Arc::new(entries_by_validator);
+
+    info!(
+        "Grouped {} validators' history entries",
+        entries_by_validator.len()
+    );
+
+    // TODO: Add check to ensure that we don't go past the current epoch
```

Comment:
Could handle this by clamping _simulation_end_epoch_ prior to calling `rebalancing_simulation`